jetbrains-core/src/software/aws/toolkits/jetbrains/core/plugins/ToolkitUpdateManager.kt

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.plugins
+
+import com.intellij.ide.plugins.IdeaPluginDescriptor
+import com.intellij.ide.plugins.PluginManagerCore
+import com.intellij.notification.NotificationAction
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.service
+import com.intellij.openapi.extensions.PluginId
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Task
+import com.intellij.openapi.updateSettings.impl.PluginDownloader
+import com.intellij.openapi.updateSettings.impl.UpdateChecker
+import com.intellij.util.Alarm
+import com.intellij.util.concurrency.annotations.RequiresBackgroundThread
+import software.aws.toolkits.core.utils.debug
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.jetbrains.AwsToolkit
+import software.aws.toolkits.jetbrains.settings.AwsSettings
+import software.aws.toolkits.jetbrains.utils.notifyInfo
+import software.aws.toolkits.resources.message
+
+class ToolkitUpdateManager {
+    private val alarm = Alarm(Alarm.ThreadToUse.SWING_THREAD)
+
+    init {
+        runTask()
+    }
+
+    private fun runTask() {
+        if (alarm.isDisposed) return
+        scheduleUpdateTask()
+
+        if (!AwsSettings.getInstance().isAutoUpdateEnabled) return
+
+        ProgressManager.getInstance().run(
+            object : Task.Backgroundable(null, "Updating AWS Toolkit") {
+                override fun run(indicator: ProgressIndicator) {
+                    checkForUpdates(indicator)
+                }
+            }
+        )
+    }
+
+    private fun scheduleUpdateTask() {
+        alarm.addRequest({ runTask() }, UPDATE_CHECK_INTERVAL_IN_MS)
+    }
+
+    @RequiresBackgroundThread
+    fun checkForUpdates(progressIndicator: ProgressIndicator) {
+        // Note: This will need to handle exceptions and ensure thread-safety
+        try {
+            val toolkitPlugin = PluginManagerCore.getPlugin(PluginId.findId(AwsToolkit.PLUGIN_ID)) ?: return
+            if (!toolkitPlugin.isEnabled) return
+            val latestToolkitPluginDownloader = getUpdate(toolkitPlugin) ?: return
+            if (!latestToolkitPluginDownloader.prepareToInstall(progressIndicator)) return
```

Comment:
nit: can we pull these `if` block to be a util predicate instead?

Reply:
what util predicate are you referring to?